### PR TITLE
Added Extension APIs to get azure account, access token and firewall rule prompt 

### PR DESCRIPTION
**Pull request by llali**
Added APIs to mssql extension to use in SQL project extension

- New APIs to prompt for firewall rule if connection fails with firewall rule error
- New APIs to return azure account and token

### CODE REVIEW

File: src/azure/azureController.ts [hunk 1]
```diff
@@ -117,48 +117,61 @@ export class AzureController {
...
+			let azureDeviceCode = await this.createDeviceCode();
+			tenantId = tenantId ? tenantId : azureDeviceCode.getHomeTenant(account).id;
+			token = await azureDeviceCode.getAccountSecurityToken(
+				account, tenantId, settings
+			);
+		}
+		return token;
+	}
+
+	public async getTokens(profile: ConnectionProfile, accountStore: AccountStore, settings: AADResource): Promise<ConnectionProfile> {
```
Comment (Charles-Gagnon) [inline src/azure/azureController.ts:155]: `populateAccountProperties` or something seems like a better name. Please also add doc comments for these new helper functions you're adding.

File: src/extension.ts [hunk 2]
```diff
@@ -64,7 +65,12 @@ export async function activate(context: vscode.ExtensionContext): Promise<IExten
 		azureFunctions: controller.azureFunctionsService,
 		getConnectionString: (connectionUriOrDetails: string | ConnectionDetails, includePassword?: boolean, includeApplicationName?: boolean) => {
 			return controller.connectionManager.getConnectionString(connectionUriOrDetails, includePassword, includeApplicationName);
-		}
+		},
+		promptForFirewallRule: async (connectionUri: string, connectionInfo: IConnectionInfo) => {
+			const connectionProfile = new ConnectionProfile(connectionInfo);
+			return await this._connectionMgr.connectionUI.addFirewallRule(connectionUri, connectionProfile);
```
Comment (Charles-Gagnon) [inline src/extension.ts:71]: You shouldn't need to await here - you can just return directly

File: src/views/connectionUI.ts [hunk 3]
```diff
@@ -488,6 +482,22 @@ export class ConnectionUI {
 		});
 	}
 
+	public async addFirewallRule(uri: string, profile: IConnectionProfile): Promise<boolean> {
+		const self = this;
```
Comment (Charles-Gagnon) [inline src/views/connectionUI.ts:486]: This shouldn't be necessary (especially since you're using `this` in the code below already anyways...)

File: typings/vscode-mssql.d.ts [hunk 4]
```diff
@@ -61,6 +66,14 @@ declare module 'vscode-mssql' {
 		 */
 		connect(connectionInfo: IConnectionInfo, saveConnection?: boolean): Promise<string>;
 
+		/**
+		 * Prompts the user to add firewall rule if connection failed with firerule error.
+		 * @param connectionUri The URI of the connection to list the databases for.
```
Comment (Charles-Gagnon) [inline typings/vscode-mssql.d.ts:71]: This param description doesn't seem correct

File: typings/vscode-mssql.d.ts [hunk 5]
```diff
@@ -284,6 +297,107 @@ declare module 'vscode-mssql' {
 		validateStreamingJob(packageFilePath: string, createStreamingJobTsql: string): Thenable<ValidateStreamingJobResult>;
 	}
 
+	/**
+ * Represents a key that identifies an account.
```
Comment (Charles-Gagnon) [inline typings/vscode-mssql.d.ts:301]: Fix indentation

File: typings/vscode-mssql.d.ts [hunk 6]
```diff
@@ -284,6 +297,107 @@ declare module 'vscode-mssql' {
...
+	export interface Token extends AccessToken {
+		/**
+		 * TokenType
+		 */
+		tokenType: string;
+	}
+
+	export interface IAzureAccountService {
+		getAccount(): Promise<IAccount>;
+		getAccountSecurityToken(account: IAccount, tenantId: string | undefined): Thenable<Token>;
```
Comment (Charles-Gagnon) [inline typings/vscode-mssql.d.ts:398]: We should be consistent and make this return a Promise

File: yarn.lock [hunk 7]
```diff
@@ -43,23 +43,23 @@
   integrity sha1-/JQj+H/K5BigrjuNVsYV2TVj1E8=
 
 "@babel/code-frame@^7.0.0":
-  version "7.14.5"
-  resolved "https://registry.yarnpkg.com/@babel/code-frame/-/code-frame-7.14.5.tgz#23b08d740e83f49c5e59945fbf1b43e80bbf4edb"
-  integrity sha512-9pzDqyc6OLDaqe+zbACgFkb6fKMNG6CObKpnYXChRsvYGyEdc7CA2BaqeOM+vOtCS5ndmJicPJhKAwYRI6UfFw==
+  version "7.16.7"
```
Comment (Charles-Gagnon) [inline yarn.lock:46]: Please revert these changes if they aren't relevant to your PR

**Review by Charles-Gagnon — COMMENTED** [review]
@llali Did you forget to add reviewers to this?

**Comment by alanrenmsft** [comment]
@cssuh these are mostly AAD related changes, please take a look.

File: src/azure/azureController.ts [hunk 8]
```diff
@@ -117,48 +117,64 @@ export class AzureController {
 		await this.prompter.promptSingle(tenantQuestion, true);
 	}
 
-	public async getTokens(profile: ConnectionProfile, accountStore: AccountStore, settings: AADResource): Promise<ConnectionProfile> {
+	public async getAccount(accountStore: AccountStore): Promise<IAccount> {
```
Comment (alanrenmsft) [inline src/azure/azureController.ts:120]: 1. The method is actually doing add account operation, right? if so we should rename the method.
2. it isweird that I have to pass in the `accountStore` object, add account to it and then return the account. the AccountStore should be globally accessible.

File: typings/vscode-mssql.d.ts [hunk 9]
```diff
@@ -61,6 +66,14 @@ declare module 'vscode-mssql' {
 		 */
 		connect(connectionInfo: IConnectionInfo, saveConnection?: boolean): Promise<string>;
 
+		/**
+		 * Prompts the user to add firewall rule if connection failed with firerule error.
+		 * @param connectionUri The URI of the connection to add firewall rule to.
+		 * @param connectionInfo The connection info
+		 * @returns True if firewall rulle added
```
Comment (Charles-Gagnon) [inline typings/vscode-mssql.d.ts:73]: rule

File: src/azure/azureController.ts (hunk 8 above)
Comment (llali) [reply src/azure/azureController.ts:120]: this method was already in the code I just divided to two methods. The previous one would add account and generate token. Now there are methods one add Account and other generates token

File: typings/vscode-mssql.d.ts [hunk 10]
```diff
@@ -61,6 +66,14 @@ declare module 'vscode-mssql' {
 		 */
 		connect(connectionInfo: IConnectionInfo, saveConnection?: boolean): Promise<string>;
 
+		/**
+		 * Prompts the user to add firewall rule if connection failed with firerule error.
```
Comment (Charles-Gagnon) [inline typings/vscode-mssql.d.ts:70]: `failed with a firewall error`

File: typings/vscode-mssql.d.ts [hunk 11]
```diff
@@ -292,6 +305,136 @@ declare module 'vscode-mssql' {
...
+		 */
+		key: IAccountKey;
+		/**
+		 * Display information for the account
+		 */
+		displayInfo: IAccountDisplayInfo;
+		/**
+		 * Custom properties stored with the account
+		 */
+		properties: any;
```
Comment (Charles-Gagnon) [inline typings/vscode-mssql.d.ts:387]: Can this be given a better type - even just an object with string key/value pairs?

File: src/extension.ts [hunk 12]
```diff
@@ -14,9 +14,17 @@ import { Deferred } from './protocol';
 import * as utils from './models/utils';
 import { ObjectExplorerUtils } from './objectExplorer/objectExplorerUtils';
 import SqlToolsServerClient from './languageservice/serviceclient';
+import { ConnectionProfile } from './models/connectionProfile';
 
 let controller: MainController = undefined;
 
+class FirewallRuleError extends Error {
```
Comment (Charles-Gagnon) [inline src/extension.ts:21]: This should go somewhere else - it's own file or generic utils file or something

File: src/azure/azureController.ts [hunk 13]
```diff
@@ -117,48 +117,64 @@ export class AzureController {
...
-			let tid = profile.tenantId ? profile.tenantId : azureDeviceCode.getHomeTenant(account).id;
-			const token = await azureDeviceCode.getAccountSecurityToken(
-				account, tid, settings
+		}
+
+		return account;
+	}
+
+	public async getAccountSecurityToken(account: IAccount, tenantId: string | undefined, settings: AADResource): Promise<Token> {
+		let token;
```
Comment (Charles-Gagnon) [inline src/azure/azureController.ts:137]: Give this a type

**Review by Charles-Gagnon — APPROVED** [review]

**Comment by Charles-Gagnon** [comment]
I assume you've done plenty of testing of AAD connections with these changes?

**Comment by llali** [comment]
> I assume you've done plenty of testing of AAD connections with these changes?

these changes won't affect anything in AAD connection. There are just APIs so other extensions can get azure account and generate token

File: src/extension.ts (hunk 12 above)
Comment (Charles-Gagnon) [reply src/extension.ts:21]: And actually - it would be good to add an interface for this in the extension typings file so that others using this can catch and handle these errors more easily.

File: src/azure/azureController.ts [hunk 14]
```diff
@@ -117,48 +117,64 @@ export class AzureController {
...
+			await this.promptForTenantChoice(account, profile);
+		}
+
+		const token = await this.getAccountSecurityToken(
+			account, profile.tenantId, settings
+		);
+
+		if (!token) {
+			let errorMessage = LocalizedConstants.msgGetTokenFail;
+			this._vscodeWrapper.showErrorMessage(errorMessage);
```
Comment (alanrenmsft) [inline src/azure/azureController.ts:171]: we need to throw exception and stop the execution here, right?

**Comment by cssuh** [comment]
I agree that we should probably put this through a full testing run before merging

**Comment by llali** [comment]
> I agree that we should probably put this through a full testing run before merging

I did test connecting with AAD auth but if you send me other scenarios, I'll do a full testing. I assume the automated tests won't cover the scenarios?

**Comment by cssuh** [comment]
Here are the 5 things to test:
- Add azure account
- Remove azure account
- Connect to azure DB (new connection)
- Connect to azure DB (existing connection)
- Add firewall rule

I can also help with the testing if needed

**Comment by Charles-Gagnon** [comment]
@cssuh Are these added to the release test plan? It would be great to use that (and add anything that's missing if it's not completed) so that we have an easy reference for these in the future.

**Comment by cssuh** [comment]
> @cssuh Are these added to the release test plan? It would be great to use that (and add anything that's missing if it's not completed) so that we have an easy reference for these in the future.

Yes I got these from the vscode-mssql ADO test plan

File: src/services/azureAccountServie.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
+/* --------------------------------------------------------------------------------------------
```
Comment (abist) [inline src/services/azureAccountServie.ts:1]: File name typo

File: typings/vscode-mssql.d.ts (hunk 11 above)
Comment (llali) [reply typings/vscode-mssql.d.ts:387]: I'm worried it breaks some scenarios for AAD.

File: src/extension.ts (hunk 12 above)
Comment (llali) [reply src/extension.ts:21]: I already added interface. will add the "implements"

File: src/azure/azureController.ts [hunk 16]
```diff
@@ -117,48 +117,64 @@ export class AzureController {
...
+		return token;
+	}
+
+	/**
+	 * Gets the token for given account and updates the connection profile with token infromation needed for AAD authetnication
+	 */
+	public async populateAccountProperties(profile: ConnectionProfile, accountStore: AccountStore, settings: AADResource): Promise<ConnectionProfile> {
+		let account = await this.getAccount(accountStore);
+
+		if (!profile.tenantId) {
```
[the remark this answers is not in the document]
Comment (llali) [reply src/azure/azureController.ts:161]: this is again existing logic but I don't mind adding a check and error message. should we prompt for tenant in that case. @cssuh please let me know what you think

[the remark this answers is not in the document]
Comment (Charles-Gagnon) [reply src/azure/azureController.ts:161]: Let's do any changes in a separate PR please so it can be cleanly reviewed.

**Review by abist — APPROVED** [review]

**Review by alanrenmsft — APPROVED** [review]

File: src/azure/azureController.ts [hunk 17]
```diff
@@ -117,48 +117,65 @@ export class AzureController {
...
-			);
-			if (!token) {
-				let errorMessage = LocalizedConstants.msgGetTokenFail;
-				this._vscodeWrapper.showErrorMessage(errorMessage);
-			}
-			profile.azureAccountToken = token.token;
-			profile.expiresOn = token.expiresOn;
-			profile.email = account.displayInfo.email;
-			profile.accountId = account.key.id;
 		} else if (config === utils.azureAuthTypeToString(AzureAuthType.DeviceCode)) {
```
Comment (llali) [inline src/azure/azureController.ts:127]: @cssuh before my change when config was AzureAuthType.DeviceCode, the getToken method didn't set the account and token in the profile. Should I keep it that way or was it a bug?

Comment (cssuh) [reply src/azure/azureController.ts:127]: hmm, that seems like a bug.  I think it should be set.

Comment (llali) [reply src/azure/azureController.ts:127]: ok I keep the change then

**Comment by llali** [comment]
> Here are the 5 things to test:
> 
> * Add azure account
> * Remove azure account
> * Connect to azure DB (new connection)
> * Connect to azure DB (existing connection)
> * Add firewall rule
> 
> I can also help with the testing if needed

@cssuh I verified all these scenarios still work as expected. The only difference I see if, when I login to my AAD account, I see a prompt for selecting tenant. I think it's expected since I have two tenants. but in existing version of mssql, I don't see any prompt using the same account. I don't know if it's a bug in the current version and fixed in my change or if I broke anything. Can you please take look at the code and let me know if you see anything wrong? Thanks

**Comment by Charles-Gagnon** [comment]
@llali That's expected - that was a feature added since the last release : https://github.com/microsoft/vscode-mssql/pull/17266

So great to hear that we have additional validation that it's working!
